### PR TITLE
Generators as Data and logp rescaling

### PR DESCRIPTION
**Pull request by ferrine**
I've implemented density rescaling and generators support

Usage
------
```python
def gen():                         
    i = 0                           
    while True:                     
        yield np.ones((20, 100)) * i
        i += 1                      
with pm.Model() as model:
    Normal('n', observed=gen(), total_size=10000)
```

### CODE REVIEW

**Comment by ferrine** [comment]
I've closed previous PR as it looked strange. I'm welcome for suggestions

**Comment by ferrine** [comment]
In my opinion it is ready for merge, interface is straightforward and tests are passing

**Comment by ferrine** [comment]
Setting new generator can be done in two steps:
```python
from pymc3.theanof import generator
def gen():                         
    i = 0                           
    while True:                     
        yield np.ones((20, 100)) * i
        i += 1      
gen_var = generator(gen())
with pm.Model() as model:
    Normal('n', observed=gen_var, total_size=10000)
gen_var.set_gen(gen())
```

File: pymc3/data.py [hunk 1]
```diff
@@ -19,3 +22,30 @@ def get_data_file(pkg, path):
...
+        self.test_value = next(generator)
+        self.gen = itertools.chain([self.test_value], generator)
+        self.tensortype = tt.TensorType(
+            self.test_value.dtype,
+            ((False, ) * self.test_value.ndim))
+
+    def __next__(self):
+        return next(self.gen)
+
+    next = __next__
```
Comment (ColCarroll) [inline pymc3/data.py]: what does this do?

Comment (ferrine) [reply pymc3/data.py]: it is for python2, generators there have `next` instead of `__next__`

File: pymc3/model.py [hunk 2]
```diff
@@ -844,14 +872,23 @@ def __init__(self, name, data, distribution, model):
...
                      for name, data in data.items()}
 
         self.missing_values = [datum.missing_values for datum in self.data.values()
                                if datum.missing_values is not None]
-        self.logp_elemwiset = distribution.logp(**self.data)
+        logp_elemwiset = distribution.logp(**self.data)
+        if total_size is None:
+            coef = tt.constant(1)
+        else:
+            assert logp_elemwiset.ndim >= 1, ('Variable with scaled density '
```
Comment (ColCarroll) [inline pymc3/model.py]: should raise an honest error, and best to raise a custom one (`assert` statements can be turned off at run time)

File: pymc3/model.py [hunk 3]
```diff
@@ -736,7 +745,14 @@ def __init__(self, type=None, owner=None, index=None, name=None,
             self.distribution = distribution
             self.tag.test_value = np.ones(
                 distribution.shape, distribution.dtype) * distribution.default()
-            self.logp_elemwiset = distribution.logp(self)
+            logp_elemwiset = distribution.logp(self)
+            if total_size is None:
+                coef = tt.constant(1)
+            else:
+                assert logp_elemwiset.ndim >= 1, ('Variable with scaled density '
```
Comment (ColCarroll) [inline pymc3/model.py]: same comment with asserts

File: pymc3/model.py [hunk 4]
```diff
@@ -814,7 +835,14 @@ def __init__(self, type=None, owner=None, index=None, name=None, data=None,
             data = as_tensor(data, name, model, distribution)
             self.missing_values = data.missing_values
 
-            self.logp_elemwiset = distribution.logp(data)
+            logp_elemwiset = distribution.logp(data)
+            if total_size is None:
+                coef = tt.constant(1)
+            else:
+                assert logp_elemwiset.ndim >= 1, ('Variable with scaled density '
```
Comment (ColCarroll) [inline pymc3/model.py]: same comment with asserts

**Comment by ferrine** [comment]
I'm starting to build OPVI using this PR, waiting for feedback. If ok - I'll merge

File: pymc3/model.py [hunk 5]
```diff
@@ -500,7 +505,8 @@ def Var(self, name, dist, data=None):
                     self.named_vars[v.name] = v
         else:
             var = ObservedRV(name=name, data=data,
-                             distribution=dist, model=self)
+                             distribution=dist, model=self,
+                             total_size=total_size)
```
Comment (twiecki) [inline pymc3/model.py]: there's an implicit convenction that `model` is always the last kwarg. Not sure it matters but might as well stick to that.

File: pymc3/model.py [hunk 6]
```diff
@@ -844,14 +874,24 @@ def __init__(self, name, data, distribution, model):
...
-        self.logp_elemwiset = distribution.logp(**self.data)
+        logp_elemwiset = distribution.logp(**self.data)
+        if total_size is None:
+            coef = tt.constant(1)
+        else:
+            if not logp_elemwiset.ndim >= 1:
+                raise ValueError('Variable with scaled density '
+                                 'needs to be at least 1 dimensional')
+            coef = tt.as_tensor(total_size) / logp_elemwiset.shape[0]
+        self.logp_elemwiset = logp_elemwiset * coef
```
Comment (twiecki) [inline pymc3/model.py]: can this be moved into `Factor` somehow to avoid the code duplication?

**Comment by ferrine** [comment]
Rebased on latest master
@twiecki, @taku-y I've made OPVI work with ADVI using these changes. Tomorrow I'm going to merge this PR and open a new one for OPVI.

**Comment by taku-y** [comment]
@ferrine Great! Looking forward to see OPVI.

**Pull request merged**